tree/30_generic_methods/variable_dict.cf

```diff
@@ -0,0 +1,52 @@
+#####################################################################################
+# Copyright 2013 Normation SAS
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Variable dict
+# @description Define a variable that contains key,value pairs (a dictionnary)
+#
+# @parameter variable_prefix The prefix of the variable name, use ${class_prefix} for a local variable and a static name for a global variable
+# @parameter variable_name   The variable to define, the full name will be variable_prefix.variable_name
+# @parameter value           The variable content in JSON format
+# 
+# @class_prefix variable_dict
+# @class_parameter variable_name
+# 
+# This bundle will define a class variable_dict_${variable_name}_{kept,repaired,not_ok,ok,reached}
+#
+# To use the generated variable, you must use the form ${variable_prefix.variable_name[key]} with each name replaced with the parameters of this method.
+# 
+# To create a local variable, use ${class_prefix} as a prefix so that it is garanteed to be unique to this technique instance.
+# to create a global variable, use a static name that you can reuse in other techniques.
+# Be careful that using a global variable can lead to unpredictable content in case of multiple definition, which is implicitly the case when a technique has more than one instance (directive). 
+# Please note that only global variables are available within templates.
```

Comment:
A variable name is made with a prefix, a dot and a suffix.

If the prefix is a static name it will behave like the global variable it is.
If the prefix is ${class_prefix} it will behave like a local variable (or more like a static variable) ie there will be one per technique instance and it will be accessible almost only from there.